### DISCUSSION TITLE
How to position popper correctly when anchor tag is wrapped?

### DISCUSSION THREAD

**Discussion by greynguyen** [floating-ui/floating-ui#1270] [Q&A]
I am trying to use popper to position a popover on an anchor that is nearest to the mouse. Right now, if I have a line that overflows like this, the popover will be displayed in the wrong position. I'd really appreciate it if anyone have any suggestions on how to solve this.

<img width="679" alt="Screen Shot 2021-04-01 at 11 08 00 PM" src="https://user-images.githubusercontent.com/15851465/113376396-a24b5200-933f-11eb-88a4-419c90b07eb4.png">

**Comment by harshagarwalnyu** [comment]
When an anchor (`<a>`) wraps to multiple lines, `getBoundingClientRect()` returns the bounding box of **all lines combined**, so the popover appears at the wrong position (top-left of the full multi-line block, not near the cursor).

**Fix: use a virtual reference element with `getClientRects()`**

`element.getClientRects()` returns one `DOMRect` per visual line. Find the rect closest to the mouse position and use it as a virtual reference:

```js
import { computePosition, flip, shift } from '@floating-ui/dom';

anchor.addEventListener('mousemove', async (event) => {
  const rects = [...anchor.getClientRects()];

  // Find the rect whose vertical center is closest to the cursor Y
  const closestRect = rects.reduce((best, rect) => {
    const rectMid = rect.top + rect.height / 2;
    const bestMid = best.top + best.height / 2;
    return Math.abs(rectMid - event.clientY) < Math.abs(bestMid - event.clientY)
      ? rect
      : best;
  });

  // Virtual reference: object with getBoundingClientRect()
  const virtualRef = {
    getBoundingClientRect: () => closestRect,
  };

  const { x, y } = await computePosition(virtualRef, popover, {
    middleware: [flip(), shift({ padding: 8 })],
  });

  Object.assign(popover.style, {
    left: `${x}px`,
    top: `${y}px`,
  });
});
```

**With `@floating-ui/react`:**
```jsx
const [virtualEl, setVirtualEl] = useState(null);

const { refs, floatingStyles } = useFloating({
  elements: { reference: virtualEl },
});

function handleMouseMove(e) {
  const rects = [...e.currentTarget.getClientRects()];
  const closest = rects.reduce(...); // same logic as above
  setVirtualEl({ getBoundingClientRect: () => closest });
}
```

This correctly positions the popover next to the specific line segment the cursor is on, regardless of how the text wraps.
